File: solutions/_144_binary_tree_preorder_traversal/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function (root) {
  if (!root) return [];
  const stack = [root];
  const res = [];

  while (stack.length) {
    const node = stack.pop();
    res.push(node.val);
    if (node.right) {
      stack.push(node.right);
    }
    if (node.left) {
      stack.push(node.left);
    }
  }

  return res;
};

module.exports = { preorderTraversal, TreeNode };

/*
https://leetcode.com/problems/binary-tree-preorder-traversal/
*/

File: solutions/_144_binary_tree_preorder_traversal/Solution.java
package _144_binary_tree_preorder_traversal;

import java.util.*;

class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if (root == null) {
            return res;
        }
        var stack = new Stack<TreeNode>();
        stack.push(root);

        while (!stack.isEmpty()) {
            var node = stack.pop();
            res.add(node.val);
            if (node.right != null) {
                stack.push(node.right);
            }
            if (node.left != null) {
                stack.push(node.left);
            }
        }
        return res;
    }
}

/*
 * https://leetcode.com/problems/binary-tree-preorder-traversal/
 */
